crates/ty_python_semantic/src/dunder_all.rs

```diff
@@ -0,0 +1,470 @@
+use rustc_hash::FxHashSet;
+
+use ruff_db::files::File;
+use ruff_db::parsed::parsed_module;
+use ruff_python_ast::name::Name;
+use ruff_python_ast::statement_visitor::{walk_stmt, StatementVisitor};
+use ruff_python_ast::{self as ast};
+
+use crate::semantic_index::ast_ids::{HasScopedExpressionId, HasScopedUseId};
+use crate::semantic_index::symbol::ScopeId;
+use crate::semantic_index::{global_scope, semantic_index, SemanticIndex};
+use crate::symbol::{symbol_from_bindings, Boundness, Symbol};
+use crate::types::{infer_expression_types, Truthiness};
+use crate::{resolve_module, Db, ModuleName};
+
+#[allow(clippy::ref_option)]
+fn dunder_all_names_cycle_recover(
+    _db: &dyn Db,
+    _value: &Option<FxHashSet<Name>>,
+    _count: u32,
+    _file: File,
+) -> salsa::CycleRecoveryAction<Option<FxHashSet<Name>>> {
+    salsa::CycleRecoveryAction::Iterate
+}
+
+fn dunder_all_names_cycle_initial(_db: &dyn Db, _file: File) -> Option<FxHashSet<Name>> {
+    None
+}
+
+/// Returns a set of names in the `__all__` variable for `file`, [`None`] if it is not defined or
+/// if it contains invalid elements.
+pub(crate) fn dunder_all_names(db: &dyn Db, file: File) -> Option<&FxHashSet<Name>> {
+    #[allow(clippy::ref_option)]
+    #[salsa::tracked(return_ref, cycle_fn=dunder_all_names_cycle_recover, cycle_initial=dunder_all_names_cycle_initial)]
+    fn dunder_all_names_impl(db: &dyn Db, file: File) -> Option<FxHashSet<Name>> {
+        let _span = tracing::trace_span!("dunder_all_names", file=?file.path(db)).entered();
+
+        let module = parsed_module(db.upcast(), file);
+        let index = semantic_index(db, file);
+        let mut collector = DunderAllNamesCollector::new(db, file, index);
+        collector.visit_body(module.suite());
+        collector.into_names()
+    }
+
+    dunder_all_names_impl(db, file).as_ref()
+}
+
+/// A visitor that collects the names in the `__all__` variable of a module.
+struct DunderAllNamesCollector<'db> {
+    db: &'db dyn Db,
+    file: File,
+
+    /// The scope in which the `__all__` names are being collected from.
+    ///
+    /// This is always going to be the global scope of the module.
+    scope: ScopeId<'db>,
+
+    /// The semantic index for the module.
+    index: &'db SemanticIndex<'db>,
+
+    /// The origin of the `__all__` variable in the current module, [`None`] if it is not defined.
+    origin: Option<DunderAllOrigin>,
+
+    /// A flag indicating whether the module uses unrecognized `__all__` idioms or there are any
+    /// invalid elements in `__all__`.
+    invalid: bool,
```

Comment:
This is mainly to reduce false positives because there are many different idioms that are being used in the ecosystem that dynamically generates the values of `__all__` in the module.